include/picongpu/simulation/stage/FieldBackground.hpp

```diff
@@ -38,55 +45,255 @@ namespace picongpu
     {
         namespace stage
         {
+            namespace detail
+            {
+                /* Implementation of background application to the given field
+                 *
+                 * @tparam T_Field field affected, e.g. picongpu::FieldE
+                 * @tparam T_FieldBackground field background to apply, e.g. picongpu::FieldBackgroundE
+                 */
+                template<typename T_Field, typename T_FieldBackground>
+                class ApplyFieldBackground
+                {
+                public:
+                    //! Field affected
+                    using Field = T_Field;
+
+                    //! Field background to apply
+                    using FieldBackground = T_FieldBackground;
+
+                    /** Create an object to apply the background
+                     *
+                     * @param cellDescription mapping for kernels
+                     * @param TODO
+                     */
+                    ApplyFieldBackground(MappingDesc const cellDescription, bool const duplicateField)
+                        : cellDescription(cellDescription)
+                        , duplicateField(duplicateField)
+                        , isEnabled(FieldBackground::InfluenceParticlePusher)
+                    {
+                        if(isEnabled && duplicateField)
+                        {
+                            // Allocate a duplicate field buffer and copy the values
+                            DataConnector& dc = Environment<>::get().DataConnector();
+                            auto field = dc.get<Field>(Field::getName(), true);
+                            auto const& gridBuffer = field->getGridBuffer();
+                            duplicateBuffer = pmacc::makeDeepCopy(gridBuffer.getDeviceBuffer());
+                            dc.releaseData(Field::getName());
+                        }
+                    }
+
+                    /** Add the field background in the given area
+                     *
+                     * @tparam T_area area to operate on
+                     *
+                     * @param step index of time iteration
+                     */
+                    template<uint32_t T_area>
+                    void add(uint32_t const step) const
+                    {
+                        if(!isEnabled)
+                            return;
+                        DataConnector& dc = Environment<>::get().DataConnector();
+                        auto& field = *dc.get<Field>(Field::getName(), true);
+                        // Always add, conditionally make a copy first
+                        if(duplicateField)
+                        {
+                            auto& gridBuffer = field.getGridBuffer();
+                            duplicateBuffer->copyFrom(gridBuffer.getDeviceBuffer());
+                            __getTransactionEvent().waitForFinished();
```

Comment:
There is no need to wait for the event system.

Reply:
Fixed